react-components/src/components/Architecture/SettingsButton.tsx

```diff
@@ -0,0 +1,215 @@
+/*!
+ * Copyright 2024 Cognite AS
+ */
+
+import { useCallback, useEffect, useMemo, useState, type ReactElement } from 'react';
+import {
+  Button,
+  Dropdown,
+  Menu,
+  Tooltip as CogsTooltip,
+  type IconType,
+  Slider
+} from '@cognite/cogs.js';
+import { useTranslation } from '../i18n/I18n';
+import { type BaseCommand } from '../../architecture/base/commands/BaseCommand';
+import { useRenderTarget } from '../RevealCanvas/ViewerContext';
+import {
+  getButtonType,
+  getDefaultCommand,
+  getFlexDirection,
+  getTooltipPlacement,
+  getIcon
+} from './utilities';
+import { LabelWithShortcut } from './LabelWithShortcut';
+import { type TranslateDelegate } from '../../architecture/base/utilities/TranslateKey';
+import styled from 'styled-components';
+import { SettingsCommand } from '../../architecture/base/concreteCommands/SettingsCommand';
+import { createButton } from './CommandButtons';
+import { BaseOptionCommand } from '../../architecture/base/commands/BaseOptionCommand';
+import { OptionButton } from './OptionButton';
+import { BaseSliderCommand } from '../../architecture/base/commands/BaseSliderCommand';
+
+export const SettingsButton = ({
+  inputCommand,
+  isHorizontal = false
+}: {
+  inputCommand: SettingsCommand;
+  isHorizontal: boolean;
+}): ReactElement => {
+  const renderTarget = useRenderTarget();
+  const { t } = useTranslation();
+  const command = useMemo<BaseCommand>(() => getDefaultCommand(inputCommand, renderTarget), []);
+
+  const [isOpen, setOpen] = useState<boolean>(false);
+  const [isEnabled, setEnabled] = useState<boolean>(true);
+  const [isVisible, setVisible] = useState<boolean>(true);
+  const [uniqueId, setUniqueId] = useState<number>(0);
+  const [icon, setIcon] = useState<IconType | undefined>(undefined);
+
+  const update = useCallback((command: BaseCommand) => {
+    setEnabled(command.isEnabled);
+    setVisible(command.isVisible);
+    setUniqueId(command.uniqueId);
+    setIcon(getIcon(command));
+  }, []);
```

Comment:
There are multiple `useState` hooks used, each of these triggers a re-render of the component when their state changes. If these states often change together, it might be more efficient to combine them into a single state object like below

```
const [state, setState] = useState({
    isOpen: false,
    isEnabled: true,
    isVisible: true,
    uniqueId: 0,
    icon: undefined
  });

  const update = useCallback((command: BaseCommand) => {
    setState({
      isEnabled: command.isEnabled,
      isVisible: command.isVisible,
      uniqueId: command.uniqueId,
      icon: getIcon(command)
    });
  }, []);
```

Reply:
Ok. Good to know. Not done it here. It we do this, I want to do it everywhere in my code